### PR TITLE
SI-8931 make generic signature consistent with interface list in classfiles

### PR DESCRIPTION
**Pull request by gourlaysama**
An optimization was introduced in 7a99c03 ([SI-5278](https://issues.scala-lang.org/browse/SI-5278)) to remove redundant
interfaces from the list of implemented interfaces in the bytecode.
However the same change was not propagated to the generic signature
of a class, which also contains a list of direct parent classes and
interfaces.

The JVM does not check the well-formedness of signatures at class
loading or linking (see §4.3.4 of jdk7 jvms), but other tools might
assume the number of implemented interfaces is the same whether one
asked for generic or erased interfaces.

It doesn't break reflection so nobody complained, but it does show:

``` scala
scala> val c = classOf[Tuple1[String]]
c: Class[(String,)] = class scala.Tuple1

scala> c.getInterfaces // Product is gone
res0: Array[Class[_]] = Array(interface scala.Product1, interface
scala.Serializable)

scala> c.getGenericInterfaces // Product is back!
res1: Array[java.lang.reflect.Type] = Array(scala.Product1<T1>,
interface scala.Product, interface scala.Serializable)
```

This moves the optimization to erasure, for use in emitting the generic
signature, and the backend calls into it later for the list of
interfaces.

### CODE REVIEW

**Comment by gourlaysama** [comment]
Hmm, that what I get for testing on quick without building locker...

**Comment by gourlaysama** [comment]
Here is the diff for scala-library.jar for this change: https://gist.github.com/gourlaysama/91e2d23ec1b60a9a186e (collections and tuples are the most affected). It isn't incredibly readable, but it helps.

**Comment by gourlaysama** [comment]
Review @retronym :)

**Comment by lrytz** [comment]
There's still duplication in the new backend (totally not your fault..!): https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala#L202

Can you also use `erasure.minimizeInterfaces` there?

**Comment by lrytz** [comment]
Looks good otherwise!

**Comment by gourlaysama** [comment]
@lrytz done. Thanks, I had completely overlooked that.

File: test/junit/scala/tools/nsc/backend/jvm/SignatureTest.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  def testGenericSignatureFiltering(): Unit = {
+    val List(cClass, _) = compileClasses(compiler)(
+      """
+      |case class C[T](_1: T) extends Product1[T]
+      """.stripMargin)
+
+    // Product1 extends Product, so the latter should not appear in C's signature
+    assertFalse(s"Signature should not contain Product: ${cClass.signature}",
+                cClass.signature.contains("Lscala/Product;"))
+  }
```
Comment (retronym) [inline test/junit/scala/tools/nsc/backend/jvm/SignatureTest.scala]: I would actually write this test as a run test that uses Java reflection to extract the list of generic interfaces. Furthermore, by setting up the classes in a self contained manner in the test, the intent is clearer.

``` scala
scala> trait B; trait A extends B; class C extends A with B
defined trait B
defined trait A
defined class C

scala> classOf[C].getGenericInterfaces
res2: Array[java.lang.reflect.Type] = Array(interface A)
```

In the test, you would just need:

``` scala
trait B; trait A extends B; class C extends A with B
object Test extends App {
   println(classOf[C].getGenericInterfaces.toList)
}
```

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 2]
```diff
@@ -192,15 +211,18 @@ abstract class Erasure extends AddInterfaces
...
           // java is unthrilled about seeing interfaces inherit from classes
-          val ok = parents filter (p => p.typeSymbol.isTrait || p.typeSymbol.isInterface)
+          val ok = minParents filter (p => p.typeSymbol.isTrait || p.typeSymbol.isInterface)
           // traits should always list Object.
           if (ok.isEmpty || ok.head.typeSymbol != ObjectClass) ObjectTpe :: ok
           else ok
         }
-        else parents
+        // a signature should always start with a class
+        else if (minParents.isEmpty || minParents.head.typeSymbol.asClass.isTrait) ObjectTpe :: minParents
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: I think we can further factor out insertion of `Object` into the list of parents here.

``` scala
def isInterfaceOrTrait(sym: Symbol) sym.isInterface || sym.isTrait

// a signature should always start with a class
def ensureClassAsFirstParent(tps: List[Type]) = tps match {
  case Nil => ObjectTpe :: Nil
  case head :: tail if isInterfaceOrTrait(head.typeSymbol) => ObjectTpe :: tps
  case _ => tps
}

val validParents = 
  if (isTraitSignature)
    minParents filter isInterfaceOrTrait // java is unthrilled about seeing interfaces inherit from classes
  else minParents
ensureClassAsFirstParent(validParents)
```

**Comment by gourlaysama** [comment]
PLS REBUILD/pr-scala@ced3ca8

File: test/junit/scala/tools/nsc/backend/jvm/SignatureTest.scala (hunk 1 above)
Comment (gourlaysama) [inline test/junit/scala/tools/nsc/backend/jvm/SignatureTest.scala]: done.

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 2 above)
Comment (gourlaysama) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala]: fixed; that's much more readable.

**Comment by retronym** [comment]
LGTM!

**Pull request merged**